pkg/cmd/pulumi-language-java/main.go

```diff
@@ -56,15 +57,35 @@ func main() {
 			cmdutil.Exit(err)
 		}
 	case binary != "":
-		logging.V(3).Infof("language host asked to use specific JAR: `%s`", binary)
-		cmd, err := lookupPath("java")
-		if err != nil {
-			cmdutil.Exit(err)
-		}
-		javaExec, err = newJarExecutor(cmd, binary)
-		if err != nil {
-			cmdutil.Exit(err)
+		logging.V(3).Infof("language host asked to use specific file: `%s`", binary)
+
+		suffix := strings.ToLower(filepath.Ext(binary))
+
+		switch suffix {
+		case ".jar":
+			cmd, err := lookupPath("java")
+			if err != nil {
+				cmdutil.Exit(err)
+			}
+			javaExec, err = newJarExecutor(cmd, binary)
+			if err != nil {
+				cmdutil.Exit(err)
+			}
+		case ".java", ".kt", ".groovy":
+			cmd, err := probeJBangExecutor()
```

Comment:
Looks like this pathway is for letting users set `binary:` option in the Pulumi.yaml and have a non-standard entry-point file. Sounds great!

Reply:
exactly